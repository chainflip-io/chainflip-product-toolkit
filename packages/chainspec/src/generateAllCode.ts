import * as fs from 'fs/promises';
import * as path from 'path';
import { specVersionCache } from '@/chainspec/cache';
import BaseCodeGenerator from './BaseCodeGenerator';
import BaseParser, { MetadataOpts } from './BaseParser';
import { diffSpecs } from './utils';

const generateAllCode = async (
  Parser: { new (opts: MetadataOpts): BaseParser },
  CodeGenerator: { new (opts?: { trackedItems: Set<string> }): BaseCodeGenerator },
  generatedDir: string,
) => {
  const info = await specVersionCache.read();

  const opts: MetadataOpts[] = Object.values(info).map((i) => ({ ...i, generatedDir }));

  const metadataForHashes = (
    await Promise.all(opts.map((o) => new Parser(o).fetchAndParseSpec()))
  ).sort((a, b) => a.specVersion - b.specVersion);

  let previousMetadata = {};

  for (const { metadata, specVersion } of metadataForHashes) {
    const specDir = path.join(generatedDir, `${specVersion}`);
    await fs.rm(specDir, { recursive: true, force: true });
    const { changedOrAddedEvents, changelog } = diffSpecs(previousMetadata, metadata);
    const generator = new CodeGenerator({ trackedItems: changedOrAddedEvents });
    for (const module of generator.generate(metadata)) {
      await module.writeFile(specDir, module.isCommon() ? changelog : undefined);
    }
    previousMetadata = metadata;
  }
};

export default generateAllCode;
