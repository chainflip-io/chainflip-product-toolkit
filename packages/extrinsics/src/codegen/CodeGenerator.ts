import { capitalize } from '@chainflip/utils/string';
import BaseCodeGenerator, { CodegenResult, Code, Identifier } from '@/chainspec/BaseCodeGenerator';
import {
  ArrayType,
  EnumType,
  MapType,
  OptionType,
  PrimitiveType,
  RangeType,
  ResolvedType,
  StructType,
  TupleType,
} from '@/chainspec/BaseParser';
import assert from 'assert';

class Struct extends Code {
  as: 'tuple' | 'record' = 'record';
  declarationType: 'type' | 'const' = 'type';

  constructor(private readonly fields: (readonly [string, CodegenResult])[]) {
    super(
      '',
      fields.map(([, type]) => type),
    );
  }

  asTuple() {
    this.as = 'tuple';
    return this;
  }

  override toString() {
    let open: string;
    let close: string;
    let transform = (name: string) => name;

    if (this.as === 'tuple') {
      open = '[';
      close = ']';
    } else {
      open = '{';
      close = '}';
      transform = (name) => name.replace(/[A-Z]/g, (m) => `_${m.toLowerCase()}`);
    }

    return `${open}\n${this.fields.map(([name, type]) => `${transform(name)}: ${type.toString()}`).join(',\n')}\n${close}`;
  }

  [Symbol.toStringTag]() {
    return 'Struct';
  }
}

export default class CodeGenerator extends BaseCodeGenerator {
  private getKnownIdentifier(name: string): string {
    switch (name) {
      case 'cfPrimitivesChainsAssetsAnyAsset':
        return 'ChainflipAsset';
      case 'cfChainsAddressEncodedAddress':
        return 'EncodedAddress';
      case 'cfChainsAddressForeignChainAddress':
        return 'ForeignChainAddress';
      case 'cfChainsBtcScriptPubkey':
        return 'BitcoinScriptPubKey';
      default:
        return capitalize(name);
    }
  }

  protected override generatePrimitive(def: PrimitiveType): CodegenResult {
    switch (def.name) {
      case 'u8':
      case 'u16':
      case 'u32':
        return new Code('number').asType();
      case 'u128':
      case 'U256':
      case 'AccountId32':
        return new Code('`0x${string}`').asType();
      case 'Bytes':
        return new Code('Uint8Array | `0x${string}`').asType();
      case 'H160':
        return new Code('Uint8Array').asType();
      default:
        throw new Error(`Method not implemented for primitive type: ${def.name}`);
    }
  }

  protected override generateEnum(def: EnumType): CodegenResult {
    const identifier = this.getKnownIdentifier(def.name);
    if (this.registry.types.has(identifier)) return new Identifier(identifier).asType();

    const isSimpleEnum = def.values.every(
      (value) => value.value.type === 'primitive' && value.value.name === 'null',
    );

    let members: CodegenResult[];

    if (isSimpleEnum) {
      members = def.values
        .filter((value) => !value.name.startsWith('__Unused'))
        .map((value) => new Code(`'${value.name}'`).asType());
    } else {
      members = def.values.map((value) => {
        const result = this.generateResolvedType(value.value);
        return new Code(`Record<'${value.name}', ${result.toString()}>`, [result]).asType();
      });
    }

    const code = new Code(members.join(' | '), members).asType();
    this.registry.types.set(identifier, code);
    return new Identifier(identifier).asType();
  }

  protected override generateStruct(def: StructType): CodegenResult {
    const fields = Object.entries(def.fields).map(([name, type]) => {
      const result = this.generateResolvedType(type);
      return [name, result] as const;
    });
    return new Struct(fields);
  }

  protected override generateArray(def: ArrayType): CodegenResult {
    if (def.value.type === 'primitive' && def.value.name === 'u8') {
      return new Code('Uint8Array').asType();
    }

    const type = this.generateResolvedType(def.value);

    return new Code(`${type.toString()}[]`, [type]);
  }

  protected override generateTuple(_def: TupleType): CodegenResult {
    throw new Error('Method not implemented.');
  }

  protected override generateRange(_def: RangeType): CodegenResult {
    throw new Error('Method not implemented.');
  }

  protected override generateOption(def: OptionType): CodegenResult {
    const type = this.generateResolvedType(def.value);
    return new Code(`${type.toString()} | null`, [type]);
  }

  protected override generateMap(_def: MapType): CodegenResult {
    throw new Error('Method not implemented.');
  }

  protected override generateItem(itemName: string, def: ResolvedType): CodegenResult {
    const type = this.generateResolvedType(def);
    assert(def.type === 'struct', `Expected struct, got ${def.type}`);
    assert(type instanceof Struct, `Expected code, got ${type[Symbol.toStringTag]()}`);
    return type.asTuple();
  }

  protected override getParserName(palletName: string, itemName: string): string {
    return `${palletName}.${itemName}`;
  }

  protected override getName(palletName: string, itemName: string): string {
    return `${capitalize(palletName)}${capitalize(itemName).replace(/_(\w)/g, (m, c: string) => c.toUpperCase())}`;
  }

  protected override getFileName(palletName: string, itemName: string): string {
    return itemName.replace(/_(\w)/g, (m, c: string) => c.toUpperCase());
  }
}
