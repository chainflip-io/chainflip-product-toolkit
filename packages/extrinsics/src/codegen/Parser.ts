import { type PalletMetadataV14, type SiLookupTypeId } from '@polkadot/types/interfaces';
import { type TypeDef } from '@polkadot/types/types';
import BaseParser, { type ResolvedType } from '@/chainspec/BaseParser';

export default class Parser extends BaseParser {
  protected getItems(pallet: PalletMetadataV14): { type: SiLookupTypeId } | null {
    return pallet.calls.isSome ? pallet.calls.unwrap() : null;
  }

  protected override shouldParsePallet(palletName: string): boolean {
    return palletName === 'Swapping';
  }

  protected override shouldParseItem(palletName: string, itemName: string): boolean {
    return itemName.includes('request_swap_deposit_address');
  }

  private depth = 0;

  protected override resolveType(type: TypeDef): ResolvedType {
    this.depth += 1;
    let result: ResolvedType;
    if (this.depth < 20) {
      result = super.resolveType(type);
    } else {
      result = { type: 'primitive', name: 'null' };
    }
    this.depth -= 1;
    return result;
  }
}
