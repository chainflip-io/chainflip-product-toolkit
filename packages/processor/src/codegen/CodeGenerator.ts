import BaseCodeGenerator, { CodegenResult, Code, Identifier } from '@/chainspec/BaseCodeGenerator';
import {
  isPrimitiveType,
  type ArrayType,
  type EnumType,
  type MapType,
  type OptionType,
  type PrimitiveType,
  type RangeType,
  type ResolvedType,
  type StructType,
  type TupleType,
} from './Parser';

const nameToIdentifier = (name: string): string =>
  name
    .replace(/(?:::|_)(.)/g, (_, c: string) => c.toUpperCase())
    .replace(/^./, (c) => c.toLowerCase());

const isNull = (type: ResolvedType) => type.type === 'primitive' && type.name === 'null';

const hexString = new Code(
  "z.string().refine((v): v is `0x${string}` => /^0x[\\da-f]*$/i.test(v), { message: 'Invalid hex string' })",
);
const numericString = new Code(
  "z.string().refine((v) => /^\\d+$/.test(v), { message: 'Invalid numeric string' })",
);
const numberOrHex = new Code(
  'z.union([z.number(), hexString, numericString]).transform((n) => BigInt(n))',
);
const accountId = new Code(
  'z.union([hexString, z.string().regex(/^[0-9a-f]+$/).transform<`0x${string}`>((v) => `0x${v}`)]).transform((value) => ss58.encode({ data: value, ss58Format: 2112 }))',
  [new Identifier('* as ss58', '@chainflip/utils/ss58')],
);
const simpleEnum = new Code(
  '<U extends string, T extends readonly [U, ...U[]]>(values: T) => z.object({ __kind: z.enum(values) }).transform(({ __kind }) => __kind!)',
);

const shortChainToLongChain = {
  Arb: 'Arbitrum',
  Btc: 'Bitcoin',
  Eth: 'Ethereum',
  Dot: 'Polkadot',
  Sol: 'Solana',
} as const;

const chainEnumMember = (name: keyof typeof shortChainToLongChain, transform?: string) =>
  `z.object({ __kind: z.literal('${name}'), value: hexString }).transform(({ value }) => ({
  chain: '${shortChainToLongChain[name]}' as const,
  address: ${transform ?? 'value'},
}))`;

export default class CodeGenerator extends BaseCodeGenerator {
  private generateEncodedAddressEnum(def: EnumType): Identifier {
    this.registry.types.set('hexString', hexString);

    const dependencies: Identifier[] = [];

    const unionMembers = def.values.map((v) => {
      switch (v.name) {
        case 'Arb':
        case 'Eth':
          return chainEnumMember(v.name);
        case 'Dot':
          dependencies.push(new Identifier('* as ss58', '@chainflip/utils/ss58'));
          return chainEnumMember(v.name, `ss58.encode({ data: value, ss58Format: 0 })`);
        case 'Btc':
          return chainEnumMember(v.name, `Buffer.from(value.slice(2), 'hex').toString('utf8')`);
        case 'Sol':
          dependencies.push(
            new Identifier('* as base58', '@chainflip/utils/base58'),
            new Identifier('hexToBytes', '@chainflip/utils/bytes'),
          );
          return chainEnumMember(v.name, `base58.encode(hexToBytes(value))`);
        default:
          throw new Error(`unknown chain: "${v.name}"`);
      }
    });

    const generated = `z.union([${unionMembers.join(',')}])`;

    this.registry.types.set(def.name, new Code(generated, dependencies));

    return new Identifier(def.name);
  }

  protected override generatePrimitive(def: PrimitiveType): CodegenResult {
    switch (def.name) {
      case 'i8':
      case 'u8':
      case 'i16':
      case 'u16':
      case 'i32':
      case 'u32':
      case 'Percent':
      case 'Permill':
        return new Code('z.number()');
      case 'u64':
      case 'u128':
      case 'U256':
        this.registry.types.set('numericString', numericString);
        this.registry.types.set('hexString', hexString);
        this.registry.types.set('numberOrHex', numberOrHex);
        return new Identifier('numberOrHex');
      case 'AccountId32':
        this.registry.types.set('hexString', hexString);
        this.registry.types.set('accountId', accountId);
        return new Identifier('accountId');
      case 'H160':
      case 'H256':
        this.registry.types.set('hexString', hexString);
        return new Identifier('hexString');
      case 'Call':
        // we would need to parse the calls as well
        return new Code('z.unknown()');
      case 'Bytes':
        this.registry.types.set('hexString', hexString);
        return new Identifier('hexString');
      case 'bool':
        return new Code('z.boolean()');
      case 'null':
        return new Code('z.null()');
    }

    throw new Error(`Unsupported primitive: ${def.name}`);
  }

  protected override generateEnum(def: EnumType): Identifier {
    if (this.registry.types.has(def.name)) return new Identifier(def.name);

    if (def.name === 'cfChainsAddressEncodedAddress') {
      return this.generateEncodedAddressEnum(def);
    }

    // "isSimple" means that none of the variants have any associated data, e.g.
    // `enum Foo { A, B, C }` and not `enum Foo { A(u32), B { field: u32 }, C }`
    const isSimple = def.values.every((v) => isNull(v.value));
    // sometimes there will be empty placeholders, we can ignore these
    const values = def.values.filter((n) => !n.name.startsWith('__Unused'));

    let generated: string;
    const dependencies: CodegenResult[] = [];

    if (isSimple) {
      this.registry.types.set('simpleEnum', simpleEnum);
      generated = `simpleEnum([${values.map((v) => `'${v.name}'`).join(', ')}])`;
    } else {
      const members = values.map((v) => {
        if (isNull(v.value)) {
          return `z.object({ __kind: z.literal('${v.name}') })`;
        }

        // if the struct has no name, it is not actually a struct but an enum variant w/ named fields
        if (v.value.type === 'struct' && v.value.name === undefined) {
          const value = this.generateStruct({
            ...v.value,
            additionalFields: {
              __kind: `z.literal('${v.name}')`,
            },
          });

          dependencies.push(value);

          return value.toString();
        }

        const value = this.generateResolvedType(v.value);

        dependencies.push(value);

        return `z.object({ __kind: z.literal('${v.name}'), value: ${value.toString()} })`;
      });

      if (members.length === 1) {
        [generated] = members;
      } else {
        generated = `z.union([${members.join(', ')}])`;
      }
    }

    this.registry.types.set(def.name, new Code(generated, dependencies));

    return new Identifier(def.name);
  }

  protected override generateStruct(def: StructType): CodegenResult {
    const structIdent = def.name && nameToIdentifier(def.name);

    if (structIdent && this.registry.types.has(structIdent)) {
      return new Identifier(structIdent);
    }

    const dependencies: CodegenResult[] = [];

    const code = new Code(
      `z.object({ ${Object.entries(def.additionalFields ?? {})
        .map(([key, value]) => `${key}: ${value}`)
        .concat(
          Object.entries(def.fields)
            .filter(([, value]) => !isNull(value))
            .map(([key, value]) => {
              const resolvedType = this.generateResolvedType(value);
              dependencies.push(resolvedType);

              if (resolvedType instanceof Identifier) {
                const identifier = resolvedType.toString();

                if (identifier === key) return key;

                return `${key}: ${identifier}`;
              }

              return `${key}: ${resolvedType.toString()}`;
            }),
        )
        .join(', ')} })`,
      dependencies,
    );

    if (!structIdent) return code;

    this.registry.types.set(structIdent, code);

    return new Identifier(structIdent);
  }

  protected override generateArray(def: ArrayType): CodegenResult {
    if (def.length) {
      if (isPrimitiveType(def.value) && def.value.name === 'u8') {
        this.registry.types.set('hexString', hexString);
        return new Identifier('hexString');
      }

      const resolvedType = this.generateResolvedType(def.value);

      const dependencies = Array.from({ length: def.length }, () => resolvedType);

      return new Code(`z.tuple([${dependencies.join(', ')}])`, [resolvedType]);
    }

    const resolvedType = this.generateResolvedType(def.value);
    const dependencies =
      resolvedType instanceof Identifier ? [resolvedType] : [...resolvedType.dependencies];
    return new Code(`z.array(${resolvedType.toString()})`, dependencies);
  }

  protected override generateTuple(def: TupleType): Code {
    const dependencies = def.values.map((t) => {
      const resolvedType = this.generateResolvedType(t);
      return resolvedType;
    });

    return new Code(`z.tuple([${dependencies.join(', ')}])`, dependencies);
  }

  protected override generateOption(def: OptionType): Code {
    const resolvedType = this.generateResolvedType(def.value);
    return new Code(`${resolvedType.toString()}.nullish()`, [resolvedType]);
  }

  protected override generateRange(def: RangeType): Code {
    const resolvedType = this.generateResolvedType(def.value);
    return new Code(
      `z.object({ start: ${resolvedType.toString()}, end: ${resolvedType.toString()} })`,
      [resolvedType],
    );
  }

  protected override generateMap(def: MapType): Code {
    const keyType = this.generateResolvedType(def.key);
    const valueType = this.generateResolvedType(def.value);
    return new Code(`z.array(z.tuple([${keyType.toString()}, ${valueType.toString()}]))`, [
      keyType,
      valueType,
    ]);
  }

  protected override getParserName(palletName: string, itemName: string): string {
    return `${palletName}.${itemName}`;
  }

  protected override getName(palletName: string, itemName: string): string {
    return `${palletName}::${itemName}`
      .replace(/(?:::|_)(.)/g, (_, c: string) => c.toUpperCase())
      .replace(/^./, (c) => c.toLowerCase());
  }

  protected override getGlobalImports(): string[] {
    return ["import { z } from 'zod';"];
  }
}
