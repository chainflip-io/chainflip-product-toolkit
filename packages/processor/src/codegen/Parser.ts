import { uncapitalize } from '@chainflip/utils/string';
import { TypeDefInfo } from '@polkadot/types';
import { type PalletMetadataV14, type SiLookupTypeId } from '@polkadot/types/interfaces';
import { type TypeDef } from '@polkadot/types/types';
import assert from 'assert';
import BaseParser from '@/chainspec/BaseParser';

const hasSubs = <T extends TypeDef>(type: T, count?: number): type is T & { sub: TypeDef[] } =>
  Array.isArray(type.sub) && (count === undefined || type.sub.length === count);

const hasSub = <T extends TypeDef>(type: T): type is T & { sub: TypeDef } =>
  !Array.isArray(type.sub) && type.sub !== undefined;

const isSi = <T extends TypeDef>(
  type: T,
): type is T & { info: TypeDefInfo.Si; type: `Lookup${number}` } =>
  type.info === TypeDefInfo.Si && /^Lookup\d+$/.test(type.type);

export type PrimitiveType = { type: 'primitive'; name: string };

export const isPrimitiveType = (type: ResolvedType): type is PrimitiveType =>
  type.type === 'primitive';

export type EnumType = {
  type: 'enum';
  name: string;
  values: { name: string; value: ResolvedType }[];
};

export type StructType = {
  type: 'struct';
  name?: string;
  fields: Record<string, ResolvedType>;
  additionalFields?: Record<string, string>;
};

export type MapType = {
  type: 'map';
  key: ResolvedType;
  value: ResolvedType;
};

export type ArrayType = { type: 'array'; value: ResolvedType; length?: number };

export type TupleType = { type: 'tuple'; values: ResolvedType[] };

export type OptionType = { type: 'option'; value: ResolvedType };

export type RangeType = { type: 'range'; value: ResolvedType };

export type ResolvedType =
  | PrimitiveType
  | EnumType
  | StructType
  | MapType
  | ArrayType
  | TupleType
  | OptionType
  | RangeType;

export default class Parser extends BaseParser {
  protected getItems(pallet: PalletMetadataV14): { type: SiLookupTypeId } | null {
    return pallet.events.isSome ? pallet.events.unwrap() : null;
  }

  protected resolveType(type: TypeDef): ResolvedType {
    const metadata = this.getMetadataSync();

    switch (type.info) {
      case TypeDefInfo.Enum: {
        assert(hasSubs(type));

        assert(type.lookupName, 'Enum type must have a lookupName');

        const result: EnumType = {
          type: 'enum',
          // inserts the pallet name to make it distinct, e.g. `CfPalletIngressEgress...` => `CfPalletSolanaIngressEgress...`
          name: uncapitalize(type.lookupName).replace('IngressEgress', this.getPalletName()),
          values: [],
        };

        for (const sub of type.sub) {
          result.values[sub.index!] = {
            name: sub.name!,
            value: this.resolveType(sub),
          };
        }

        return result;
      }
      case TypeDefInfo.Struct: {
        assert(hasSubs(type));

        const result: StructType = {
          type: 'struct',
          name: type.lookupName ?? this.genericNamespace(type.namespace),
          fields: {},
        };

        for (const sub of type.sub) {
          result.fields[sub.name!] = this.resolveType(sub);
        }

        return result;
      }
      case TypeDefInfo.Si:
        assert(isSi(type));
        return this.resolveType(metadata.registry.lookup.getTypeDef(type.type));
      case TypeDefInfo.Compact:
        assert(hasSub(type));
        return this.resolveType(type.sub);
      case TypeDefInfo.Null:
        return { type: 'primitive', name: 'null' };
      case TypeDefInfo.Plain:
        return { type: 'primitive', name: type.type };
      case TypeDefInfo.BTreeSet:
      case TypeDefInfo.Vec:
      case TypeDefInfo.VecFixed:
        assert(hasSub(type));
        return {
          type: 'array',
          value: this.resolveType(type.sub),
          length: type.length,
        };
      case TypeDefInfo.Tuple:
        assert(hasSubs(type));

        return {
          type: 'tuple',
          values: type.sub.map((t) => this.resolveType(t)),
        };
      case TypeDefInfo.Option:
        assert(hasSub(type));
        return {
          type: 'option',
          value: this.resolveType(type.sub),
        };
      case TypeDefInfo.Result:
        assert(hasSubs(type));
        return {
          type: 'enum',
          name: uncapitalize(type.typeName!),
          values: [
            {
              name: 'Ok',
              value: this.resolveType(type.sub[0]),
            },
            {
              name: 'Err',
              value: this.resolveType(type.sub[1]),
            },
          ],
        };
      case TypeDefInfo.Range:
        assert(hasSub(type));
        return {
          type: 'range',
          value: this.resolveType(type.sub),
        };
      case TypeDefInfo.BTreeMap:
        assert(hasSubs(type, 2));

        return {
          type: 'map',
          key: this.resolveType(type.sub[0]),
          value: this.resolveType(type.sub[1]),
        };
      default:
        console.error(type);
        throw new Error(`Unhandled type: ${type.info}`);
    }
  }
}
