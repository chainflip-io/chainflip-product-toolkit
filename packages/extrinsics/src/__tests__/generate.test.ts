import * as fs from 'fs/promises';
import * as os from 'os';
import * as path from 'path';
import { describe, it, expect } from 'vitest';
import generateAllCode from '@/chainspec/generateAllCode';
import CodeGenerator from '../codegen/CodeGenerator';
import Parser from '../codegen/Parser';

async function* readdir(dir: string): AsyncGenerator<string> {
  const dirents = await fs.readdir(dir, { withFileTypes: true });

  for (const dirent of dirents) {
    const res = path.resolve(dir, dirent.name);
    if (dirent.isDirectory()) {
      yield* readdir(res);
    } else if (dirent.isFile()) {
      yield res;
    }
  }
}

describe('extrinsic codegen', () => {
  it('generates the code in the desired format', async () => {
    const dir = await fs.mkdtemp(path.join(os.tmpdir(), 'extrinsics-codegen-'));

    await generateAllCode(Parser, CodeGenerator, dir);

    for await (const file of readdir(dir)) {
      if (!file.includes('/160/')) continue;
      const content = await fs.readFile(file, 'utf-8');
      expect(content).toMatchSnapshot(file.replace(dir, ''));
    }
  });
});
