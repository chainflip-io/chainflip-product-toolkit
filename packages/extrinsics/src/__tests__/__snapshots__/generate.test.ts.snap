// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extrinsic codegen > generates the code in the desired format > /160/CHANGELOG.md 1`] = `
"Swapping:
  - request_swap_deposit_address:
    - sourceAsset: added
    - destinationAsset: added
  - request_swap_deposit_address_with_affiliates:
    - dcaParameters: added
    - sourceAsset: added
    - destinationAsset: added
    - refundParameters: changed
"
`;

exports[`extrinsic codegen > generates the code in the desired format > /160/common.ts 1`] = `
"export type ChainflipAsset =
  | 'Eth'
  | 'Flip'
  | 'Usdc'
  | 'Dot'
  | 'Btc'
  | 'ArbEth'
  | 'ArbUsdc'
  | 'Usdt'
  | 'Sol'
  | 'SolUsdc';

export type EncodedAddress =
  | Record<'Eth', Uint8Array>
  | Record<'Dot', Uint8Array>
  | Record<'Btc', Uint8Array | \`0x\${string}\`>
  | Record<'Arb', Uint8Array>
  | Record<'Sol', Uint8Array>;

export type BitcoinScriptPubKey =
  | Record<'P2PKH', Uint8Array>
  | Record<'P2SH', Uint8Array>
  | Record<'P2WPKH', Uint8Array>
  | Record<'P2WSH', Uint8Array>
  | Record<'Taproot', Uint8Array>
  | Record<
      'OtherSegwit',
      {
        version: number;
        program: Uint8Array | \`0x\${string}\`;
      }
    >;

export type ForeignChainAddress =
  | Record<'Eth', Uint8Array>
  | Record<'Dot', Uint8Array>
  | Record<'Btc', BitcoinScriptPubKey>
  | Record<'Arb', Uint8Array>
  | Record<'Sol', Uint8Array>;
"
`;

exports[`extrinsic codegen > generates the code in the desired format > /160/swapping/requestSwapDepositAddress.ts 1`] = `
"import { type ChainflipAsset, type EncodedAddress } from '../common';

export type SwappingRequestSwapDepositAddress = [
  sourceAsset: ChainflipAsset,
  destinationAsset: ChainflipAsset,
  destinationAddress: EncodedAddress,
  brokerCommission: number,
  channelMetadata: {
    message: Uint8Array | \`0x\${string}\`;
    gas_budget: \`0x\${string}\`;
    cf_parameters: Uint8Array | \`0x\${string}\`;
  } | null,
  boostFee: number,
];
"
`;

exports[`extrinsic codegen > generates the code in the desired format > /160/swapping/requestSwapDepositAddressWithAffiliates.ts 1`] = `
"import { type ChainflipAsset, type EncodedAddress, type ForeignChainAddress } from '../common';

export type SwappingRequestSwapDepositAddressWithAffiliates = [
  sourceAsset: ChainflipAsset,
  destinationAsset: ChainflipAsset,
  destinationAddress: EncodedAddress,
  brokerCommission: number,
  channelMetadata: {
    message: Uint8Array | \`0x\${string}\`;
    gas_budget: \`0x\${string}\`;
    cf_parameters: Uint8Array | \`0x\${string}\`;
  } | null,
  boostFee: number,
  affiliateFees: {
    account: \`0x\${string}\`;
    bps: number;
  }[],
  refundParameters: {
    retry_duration: number;
    refund_address: ForeignChainAddress;
    min_price: \`0x\${string}\`;
  } | null,
  dcaParameters: {
    number_of_chunks: number;
    chunk_interval: number;
  } | null,
];
"
`;
